services/movie_session.py

```
@@ -0,0 +1,71 @@
+from django.db.models import QuerySet
+from db.models import MovieSession
+from datetime import datetime
+from typing import Optional
+
+
+def create_movie_session(
+        movie_show_time: datetime,
+        movie_id: int,
+        cinema_hall_id: int
+) -> MovieSession:
+    return MovieSession.objects.create(
+        show_time=movie_show_time,
+        movie_id=movie_id,
+        cinema_hall_id=cinema_hall_id
+    )
+
+
+def get_movie_sessions(
+        sessions_date: Optional[str] = None
+) -> QuerySet:
+    sessions = MovieSession.objects.all()
+    if sessions_date:
+        try:
+            parsed_date = datetime.strptime(
+                sessions_date,
+                "%Y-%m-%d"
+            ).date()
+            sessions = sessions.filter(
+                show_time__date=parsed_date
+            )
+        except ValueError:
+            return MovieSession.objects.none()
+    return sessions
+
+
+def get_movie_session_by_id(
+        movie_session_id: int
+) -> Optional[MovieSession]:
+    return MovieSession.objects.filter(id=movie_session_id).first()

```

Single-object retrieval should use `.get()` instead of `.filter(...).first()`. The checklist states: "Don't use `.filter()` when you need to get one item, use `.get()` instead." Replace `MovieSession.objects.filter(id=movie_session_id).first()` with `MovieSession.objects.get(id=movie_session_id)` (and handle DoesNotExist if you want to return None or let the exception propagate). See checklist .